capa/features/extractors/vmray/models.py

```diff
@@ -0,0 +1,292 @@
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+from typing import Dict, List, Union, Optional
+
+from pydantic import Field, BaseModel
+from typing_extensions import Annotated
+from pydantic.functional_validators import BeforeValidator
+
+"""
+# possible param types, included for documentation
+PARAM_TYPE = (
+    "signed_8bit",
+    "unsigned_8bit",
+    "signed_16bit",
+    "unsigned_16bit",
+    "signed_32bit",
+    "unsigned_32bit",
+    "signed_64bit",
+    "unsigned_64bit",
+    "double",
+    "void_ptr",
+    "bool",
+    "unknown",
+    "ptr",
+    "void",
+    "str",
+    "array",
+    "container",
+    "bindata",
+    "undefined_type",
+)
+"""
+
+PARAM_TYPE_PTR = ("void_ptr", "ptr")
+PARAM_TYPE_STR = ("str",)
+PARAM_TYPE_INT = (
+    "signed_8bit",
+    "unsigned_8bit",
+    "signed_16bit",
+    "unsigned_16bit",
+    "signed_32bit",
+    "unsigned_32bit",
+    "signed_64bit",
+    "unsigned_64bit",
+    "double",
+    "bool",
+    "unknown",
+)
+
+
+def hexint(value: Union[str, int]) -> int:
+    if isinstance(value, str):
+        return int(value, 16) if value.startswith("0x") else int(value, 10)
+    else:
+        return value
+
+
+def validate_hex_int(value: Union[str, int]) -> int:
+    return hexint(value)
+
+
+def validate_param_list(value):
+    if isinstance(value, list):
+        return value
+    else:
+        return [value]
+
+
+# convert the input value to a Python int type before inner validation (int) is called
+HexInt = Annotated[int, BeforeValidator(validate_hex_int)]
+
+
+class ParamDeref(BaseModel):
+    type_: str = Field(alias="type")
+    value: Optional[str] = None
+
+
+class Param(BaseModel):
+    name: str
+    type_: str = Field(alias="type")
+    value: Optional[str] = None
+    deref: Optional[ParamDeref] = None
+
+
+# params may be stored as a list of Param or a single Param so we convert
+# the input value to Python list type before the inner validation (List[Param])
+# is called
+ParamList = Annotated[List[Param], BeforeValidator(validate_param_list)]
+
+
+class Params(BaseModel):
+    params: ParamList = Field(alias="param")
+
+
+# models flog.xml files
+class FunctionCall(BaseModel):
+    # ts: HexInt
+    fncall_id: HexInt
+    process_id: HexInt
+    thread_id: HexInt
+    name: str
+    # addr: HexInt
+    # from_addr: HexInt = Field(alias="from")
+    params_in: Params = Field(alias="in", default=None)
+    # params_out: Params = Field(alias="out", default=None)
+
+
+"""
+# not useful for capa, but included for documentation in case
+class FunctionReturn(BaseModel):
+    ts: HexInt
+    fncall_id: HexInt
+    addr: HexInt
+    from_addr: HexInt = Field(alias="from")
+"""
+
+
+class Analysis(BaseModel):
+    log_version: str  # tested 2
```

Comment:
i think this should probably be validated as ` = "2"` so that we don't try to interpet log files that we don't support.

Reply:
https://github.com/mandiant/capa/pull/2208/commits/b68a91e10b68a84405fe05de787d110df6cad731